src/sound/loopysound.cpp

```diff
@@ -0,0 +1,901 @@
+			if(midi_param_count >= message_size && !midi_in_sysex)
+			{
+				if(channels[channel].midi_enabled)
+				{
+					switch(status_hi)
+					{
+					case 0x8:
+						note_off(channel, midi_param_bytes[0]);
+						break;
+					case 0x9:
+						if(midi_param_bytes[1] > 0) note_on(channel, midi_param_bytes[0]);
+						else note_off(channel, midi_param_bytes[0]);
+						break;
+					case 0xA:
+						printf("[Sound] unhandled message KEY PRESSURE\n");
+						break;
+					case 0xB:
+						if(midi_param_bytes[0] == 0x40)
+						{
+							control_chg_sustain(channel, (midi_param_bytes[1] >= 0x40));
+						}
+						else
+						{
+							printf("[Sound] unhandled message CONTROL CHANGE %02X %02X\n", midi_param_bytes[0], midi_param_bytes[1]);
+						}
+						break;
+					case 0xC:
+						prog_chg(channel, midi_param_bytes[0]);
+						break;
+					case 0xD:
+						printf("[Sound] unhandled message CHANNEL PRESSURE\n");
+						break;
+					case 0xE:
+						pitch_bend(channel, (midi_param_bytes[1]<<1) | (midi_param_bytes[1]>>6));
+						break;
+					case 0xF:
+					default:
+						break;
+					}
+				}
+				midi_param_count = 0;
+				midi_status = midi_running_status;
+			}
+		}
+	}
+}
+
+int UPD937_Core::read_rom_8(int offset)
+{
+	return rom[offset&rom_mask]&0xFF;
+}
+
+int UPD937_Core::read_rom_16(int offset)
+{
+	return ((rom[(offset+1)&rom_mask]&0xFF)<<8)|(rom[offset&rom_mask]&0xFF);
+}
+
+int UPD937_Core::read_rom_24(int offset)
+{
+	return ((rom[(offset+2)&rom_mask]&0xFF)<<16)|((rom[(offset+1)&rom_mask]&0xFF)<<8)|(rom[offset&rom_mask]&0xFF);
+}
+
+void UPD937_Core::update_sample()
+{
+	// Clock the volume & pitch envelope generators
+	if((sample_count%384) == 0) update_volume_envelopes();
+	int clk2_div = (int) round(CLK2_DIVP * synthesis_rate);
+	clk2_counter += CLK2_MUL;
+	if(clk2_counter >= clk2_div)
+	{
+		update_pitch_envelopes();
+		clk2_counter -= clk2_div;
+	}
+
+	// Update volume/pitch ramps
+	for(int v = 0; v < 32; v++)
+	{
+		UPD937_VoiceState *vo = &voices[v];
+		vo->volume_rate_counter++;
+		if(vo->volume_rate_counter >= vo->volume_rate_div)
+		{
+			vo->volume_rate_counter = 0;
+			if(vo->volume_down)
+			{
+				vo->volume = std::clamp(std::max(vo->volume_target, vo->volume - vo->volume_rate_mul), 0, 65535);
+			}
+			else
+			{
+				vo->volume = std::clamp(std::min(vo->volume_target, vo->volume + vo->volume_rate_mul), 0, 65535);
+			}
+		}
+		if(vo->volume > 0)
+		{
+			int pitch_relative = vo->pitch;
+			pitch_relative += vo->pitch_env_value/16;
+			pitch_relative += channels[vo->channel].bend_offset;
+			//if(pitch_relative < 0) pitch_relative = 0;
+			//if(pitch_relative > 0x5FF) pitch_relative = 0x5FF;
+			vo->sample_fract += read_rom_16(ptr_pitchtable + pitch_relative*2);
+			if(vo->sample_fract >= 0x8000)
+			{
+				vo->sample_fract -= 0x8000;
+				vo->sample_last_val = (read_rom_16(vo->sample_ptr*2)>>4) - 0x800;
+				vo->sample_ptr++;
+			}
+			if(vo->sample_ptr > vo->sample_end) vo->sample_ptr = vo->sample_loop;
```

Comment:
Just checking - should this be `>=` instead? I would expect the end address to be exclusive, but perhaps that's not the case here

Reply:
End address is inclusive, testing with exclusive is horribly detuned as expected.